Allow moving meshes without motion vectors

Ansraer opened the pull request:
This PR makes it possible to translate meshes without writing their movement to the movement buffer.

Right now, this PR integrates with the physics_interpolation feature to handle the propagation of "teleports" through the scene. To teleport something, the user simply has to update the transform and then call `reset_physics_interpolation`.

Alternatively the user can use the optional "teleport" parameter when calling manually calling the RenderingServer's bound transform method.

<details>
<summary>Video Example</summary>

https://github.com/user-attachments/assets/524b334b-c0e2-4a33-83e5-deaf6180d807
</details>


- Supersedes #77523
- Fixes https://github.com/TokisanGames/Terrain3D/issues/302

TokisanGames commented:
This works very well to resolve issues with FSR 1, 2.2, and TAA on Terrain3D. All three look good without any artifact issues. Motion vectors are sane. At least this is the case on my RTX 3070.
Thank you very much!

bruvzg commented on doc/classes/MultiMesh.xml:
```diff
@@ -94,6 +94,7 @@
 			<return type="void" />
 			<param index="0" name="instance" type="int" />
 			<param index="1" name="transform" type="Transform3D" />
+			<param index="2" name="teleport" type="bool" default="false" />
```
Adding new arguments to a public method require binding compatibility methods for GDExtension, see https://docs.godotengine.org/en/stable/contributing/development/handling_compatibility_breakages.html

Ansraer replied:
Already fixed.
Sorry about missing that, it would appear my git pre commit hooks are outdated. Been a while since I had enough time to upstream something.

mrjustaguy commented:
FSR 1 has no temporal components, it's purely spatial.

lawnjelly commented:
Given the now confusing naming, an option we could also consider longterm is adding an alias `teleport()` function instead of `reset_physics_interpolation()` (and perhaps deprecate the old function name but leave it available for compatibility).

`teleport()` was always my preference for terminology (and is used in smoothing addon).

Although that would be for separate proposal, as it has implications for docs etc.

Bear in mind also that this PR will need some rejigging after #104269, (or vice versa if this is merged first).

Ansraer commented:
> Given the now confusing naming, an option we could also consider longterm is adding an alias `teleport()` function instead of `reset_physics_interpolation()` (and perhaps deprecate the old function name but leave it available for compatibility).

I thought about that but decided it was out of scope of this PR since it would require me to propagate the teleport through the scene tree, something which right now `reset_physics_interpolation()` does for me.

I suppose an optional teleport parameter could be added to Node3D's exposed transform functions, similar to what this PR is doing internally right now.
Implementation wise the transform functions would just have to call `reset_physics_interpolation()` (maybe renamed to `reset_interpolation()`) after updating the transform if the new parameter is true.

> `teleport()` was always my preference for terminology (and is used in smoothing addon).

Yeah, I also called the optional parameter "teleport" in this PR.

> Bear in mind also that this PR will need some rejigging after #104269, (or vice versa if this is merged first).

Yeah. I considered building this PR on top of yours for a while, but in the end decided to stick to master. This is a significantly smaller change than your interpolation refactor, so I am hoping to get it merged quickly.

lawnjelly commented:
> I thought about that but decided it was out of scope of this PR since it would require me to propagate the teleport through the scene tree, something which right now reset_physics_interpolation() does for me.

Agree is out of scope, but just to clarify my suggestion was to rename the existing `reset_physics_interpolation()` functionality to `teleport()`, rather than add in extra machinery.

> I suppose an optional teleport parameter could be added to Node3D's exposed transform functions, similar to what this PR is doing internally right now.

We did consider this at the time, but a separate function was preferred in general (with the idea that it would rarely be used), and it had the benefit of being backward compatible. I tended to think this made sense from an efficiency point of view too, as `set_transform()` is used *a lot*, and there's no sense in sending an extra bool all the time if it isn't needed.

lawnjelly commented on servers/rendering_server.h:
```diff
@@ -1426,7 +1429,7 @@ class RenderingServer : public Object {
 	virtual void instance_set_layer_mask(RID p_instance, uint32_t p_mask) = 0;
 	virtual void instance_set_pivot_data(RID p_instance, float p_sorting_offset, bool p_use_aabb_center) = 0;
-	virtual void instance_set_transform(RID p_instance, const Transform3D &p_transform) = 0;
+	virtual void instance_set_transform(RID p_instance, const Transform3D &p_transform, bool p_teleported = false) = 0;
```
I'm not sure this is a good idea, as said in comment, this idea was previously rejected, and preference was for a separate server function.

lawnjelly replied:
The issue mentioned before is referring to the command queue. Ideally we want to reduce as much as possible traffic between scene side and the servers, because all these commands have to be stored to be processed later (assuming not using single threaded).

`instance_set_transform()` is a pretty hot function, it may end up getting called tens of thousands of times in a frame. Having the machinery to store 10,000 unnecessary bools in the queue is a cost, that might be say 1000,000 or more bools crossing the queue per second, totally unnecessarily.

If it were a simple c++ function with no queue in between client and server, and the ability to inline / optimize etc, then I would be in agreement.

Wrappers on the server side on the other hand will be fully optimized by the compiler, there's no problem having a `reset` function grab the instance, and call `set_transform(xform, true)` internally (if that's the simplest way to implement it your side).

lawnjelly replied:
> Ah shit

Had the same moment, many times :blush:

TokisanGames commented:
To make this PR and Physics Interpolation work in Terrain3D, we are currently doing this:
```
	RS->instance_set_transform(mesh_array[instance], t, true); // added true for this PR
	RS->instance_reset_physics_interpolation(mesh_array[instance]);
```

They are right next to each other every time we reposition the clipmap terrain. It would make sense to me to have one function call reset motion vectors for both PI and TAA/FSR. I think a name like `reset_motion_vectors` is clearer than `teleport`, but ultimately don't care as long as I have a clear, documented way to reset them, which has now been provided.

lawnjelly commented:
Although we don't always future proof, the problem with using specific names like `reset_physics_interpolation` and `reset_motion_vectors` is that it doesn't take account that typically _several_ systems need to get reset on a teleport.

The user should ideally have one function to call that does it all.

Consider that e.g. spatial audio should also be interpolated eventually, just to give an example. Granted that's a different server, but this collision of functionality has already occurred within `VisualServer`, and it's a fair bet it may happen again.

Ansraer commented:
> Although we don't always future proof, the problem with using specific names like reset_physics_interpolation and reset_motion_vectors is that it doesn't take account that typically several systems need to get reset on a teleport.

> If a user has turned off physics interpolation for a branch, the resets may not be propagated currently. We may want to alter this so that the reset (for motion vectors) still gets sent to the server. This is something that can be fixed up after https://github.com/godotengine/godot/pull/104269 though.

Honestly, I think we will need a  generic `reset_interpolation` at some point in the future. However, implementing that might be tricky. If this function is always called after the property that should be reset has already been changed I am no certain if we can guarantee that none of the other threads has already reacted to this change and interpolated it before receiving the reset command. Ahh, the joys os multithreading.

> If you aren't rebasing on top of https://github.com/godotengine/godot/pull/104269, you can probably use the existing instance_reset_physics_interpolation() function, making a lot of the changes here unnecessary I think.

Honestly not really. Most of the changes here are the plumbing needed to expose the optional parameter, something I would have done anyways since being able to turn on teleportation while setting the transform is essential imo.

You are correct that if I had patched that function I could have left VisualInstance3D as it is, but since I already have the teleport parameter I might as well use it. Performance is the same and code readability is better when we are explicit about setting transform with teleport enabled.

However, just for completeness sake I will add teleportation to `instance_reset_physics_interpolation()`. It is not really needed since imo the proper way is to already enable teleportation when setting the transform, but as long as that method exists it should behave as users expect it to. Feel free to purge it in the future.

TokisanGames commented:
What is occurring? 

It most definitely is not teleporting. The mesh isn't moving as a result of the function call. 

It is resetting the motion vector to its initial state, ie. 0. You could also say it's clearing the motion vector. 

Is it doing anything other than setting motion vector to zero?

lawnjelly commented:
> It most definitely is not teleporting. The mesh isn't moving as a result of the function call.
> It is resetting the motion vector to its initial state, ie. 0. You could also say it's clearing the motion vector.
> Is it doing anything other than setting motion vector to zero?

Sorry, this was discussed in rendering meeting, too much to repeat.

I have (for as long as I remember, and this is like over 2 decades :grin: ) used the term `teleport` but that's probably just ingrained now. `reset_interpolation` is probably fine too, but then again, is motion vectors interpolation? Motion vectors might be used for e.g. motion blur etc. You see again the problem of using a specific name.

Thinking about it, if you are dead set on proceeding before #104269, an even better option would be just adding the `teleport` / `reset_interpolation` function to the `RenderingServer` and leaving all the scene side stuff to a follow up. We do this sometimes, users can call these directly with the `RIDs` temporarily (using `VisualInstance::get_instance()` to get the `RID`).

It would enable users and yourself to test the server side functionality in the next dev build, without wasting time and energy debugging / responding to issues on scene side stuff which will be changing significantly soon, and will need to be ripped out. Otherwise users will be confused, they won't understand why it has been added out of order. Sometimes we have dependent PRs, it's a common problem. :shrug: 

I would be happy to put in the plumbing scene side for this if you go this way, BTW, as I'm very familiar with that side and the pitfalls, and it's obviously got to rig in with the existing physics interpolation in a sensible way regarding propagation of resets etc.

Ansraer commented:
> It would enable users and yourself to test the server side functionality in the next dev build, without wasting time and energy debugging / responding to issues on scene side stuff which will be changing significantly soon, and will need to be ripped out.

I myself would be totally fine with that. My own project interacts with RenderingServer directly, so I honestly don't really care too much about the scene side of things. I only added it to this PR for completeness’s sake and to have something that can be easily tested when evaluating this PR.

But if you are willing to take care of this in your PR I will modify this to only touch the RenderingServer.

I am just gonna call it `reset_interpolation`. Gimme an hour to refactor everything.

TokisanGames commented:
I'm getting this error message now w/ FSR2 enabled.
```
ERROR: Condition "p_job.uavMip[i] >= mip_slice_rids.size()" is true. Returning: FFX_ERROR_INVALID_ARGUMENT
   at: execute_gpu_job_compute_rd (servers\rendering\renderer_rd\effects\fsr2.cpp:418)
```

Also the grass particle shader now stutters with TAA. I think it's worse than before. It's also bad with PI, but I didn't test that yesterday. We do need to reset PI/MV on grass.

The ground looks good though in all modes though, except for the error.

mrjustaguy commented:
FSR 1 will not produce MV issues because it doesn't use them (it's a purely Spatial Upscaler & Sharpener), so no need to spend your time testing FSR 1

Ansraer commented:
> I'm getting this error message now w/ FSR2 enabled.
Huh, that is interesting. I don't think I changed anything that should mess with that. Made some more minor changes, if you have some time I would appreciate it if you could try again, maybe I already fixed that.

If the issue still exists just report back and I will look into it some more once I am home. Looking at my PR on a phone I can't see how I could be causing this with the current version.

Calinou commented on doc/classes/RenderingServer.xml:
```diff
@@ -1931,6 +1931,13 @@
+			<return type="void" />
+			<param index="0" name="instance" type="RID" />
+			<description>
+				Resets motion vectors and other interpolated values.
```
Does this also affect physics interpolation? This should be mentioned in the description. I would also reference `[method Node.reset_physics_interpolation]` within this description (and do the same in that method's description, reference `[method RenderingServer.instance_reset_interpolation]` there).

Ansraer replied:
Lawnjelly is currently rewriting the physics stuff and plans to remove ´reset_physics_interpolation´ entirely. Right now the main focus of this PR is to get this working only on the rendering side, physics will be added by lawnjelly in a followup PR.

Calinou commented on doc/classes/MultiMesh.xml:
```diff
@@ -106,6 +106,14 @@
+			<param index="0" name="instance" type="int" />
+			<param index="1" name="transform" type="Transform3D" />
+			<description>
+				Sets the [Transform2D] for a specific instance without writing motion vectors.
```
```suggestion
				Sets the [Transform2D] for a specific instance without writing motion vectors. Use this when teleporting a mesh from a position to another to avoid ghosting artifacts.
```

Calinou commented on doc/classes/RenderingServer.xml:
```diff
@@ -1931,6 +1931,13 @@
+			<return type="void" />
+			<param index="0" name="instance" type="RID" />
+			<description>
+				Resets motion vectors and other interpolated values.
```
```suggestion
				Resets motion vectors and other interpolated values. Use this [i]after[/i] teleporting a mesh from a position to another to avoid ghosting artifacts.
```

lawnjelly commented on servers/rendering_server.h:
```diff
@@ -466,6 +466,7 @@ class RenderingServer : public Object {
 
 	virtual void multimesh_set_mesh(RID p_multimesh, RID p_mesh) = 0;
 	virtual void multimesh_instance_set_transform(RID p_multimesh, int p_index, const Transform3D &p_transform) = 0;
+	virtual void multimesh_instance_set_transform_teleport(RID p_multimesh, int p_index, const Transform3D &p_transform) = 0;
```
This should probably more sensibly hook in with the existing `multimesh_instance_reset_physics_interpolation()` on line 494. The current transform should already be available in the server if required. When physics interpolation is in use, the (interpolated) transform won't be available anyway scene side.

_(`Multimeshes` are not changed in the new `SceneTreeFTI` BTW as there are no pressing reasons to move the multimesh instances scene side.)_

The naming isn't ideal currently, but we don't want to be calling two separate server functions here to do the same thing. If necessary we can consider renaming in a followup but it would be compat breaking.

lawnjelly commented on servers/rendering_server.h:
```diff
@@ -1434,6 +1435,8 @@ class RenderingServer : public Object {
 	virtual void instance_set_surface_override_material(RID p_instance, int p_surface, RID p_material) = 0;
 	virtual void instance_set_visible(RID p_instance, bool p_visible) = 0;
 
+	virtual void instance_reset_interpolation(RID p_instance) = 0;
```
Still not sure this name is good. With the `SceneTreeFTI`, this command will (currently anyway) no longer have any effect on the interpolation in the server.

Ansraer committed:
allow moving meshes without motion vectors

TokisanGames commented:
> FSR 1 will not produce MV issues because it doesn't use them (it's a purely Spatial Upscaler & Sharpener), so no need to spend your time testing FSR 1

The quoted error was new and related to FSR. I was confirming that the error appeared only when FSR2 was enabled, and not in other modes: disabled, FSR1, etc. Nothing to do with MV in this case.

> I don't think I changed anything that should mess with that. 

I tested more and found:
* TAA & FSR2 are very bad w/o this PR
* They're solid with this PR
* The FSR2 error exists in `master`

`ERROR: Condition "p_job.uavMip[i] >= mip_slice_rids.size()" is true. Returning: FFX_ERROR_INVALID_ARGUMENT`

lawnjelly commented on doc/classes/RenderingServer.xml:
```diff
@@ -2046,6 +2046,13 @@
+			<return type="void" />
+			<param index="0" name="instance" type="RID" />
+			<description>
+				Resets motion vectors and other interpolated values. Use this [i]after[/i] teleporting a mesh from a position to another to avoid ghosting artifacts.
```
Not a biggie, but it's slightly more expected native english to say "from one position to another" rather than "from a position to another" (both are semantically correct thought).

lawnjelly reviewed (approved):
Much better imo.

This should work fine and I can rejig calling `instance_teleport` when rebasing the `SceneTreeFTI` PR.

Probably needs a double check by a member of the 4.x rendering team as I'm not so active there, just to check that logic meets their needs with the motion vectors.

Imo the most important aspect to get right was always the server API function / s, because once set, these are difficult to change without changing backward compat. It is easy to fix bugs / make adjustments in the server later.

clayjohn reviewed (approved):
Looks good to me

AThousandShips commented on servers/rendering/renderer_scene_cull.h:
```diff
@@ -1054,6 +1055,8 @@ class RendererSceneCull : public RenderingMethod {
 	virtual void instance_set_visible(RID p_instance, bool p_visible);
 	virtual void instance_geometry_set_transparency(RID p_instance, float p_transparency);
 
+	virtual void instance_teleport(RID instance);
```
```suggestion
	virtual void instance_teleport(RID p_instance);
```
Please prefix arguments with `p_`

Repiteo commented:
Thanks!
